### PR TITLE
feat: Customizable Loops 1/n

### PR DESCRIPTION
**Pull request by aniketmaurya**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

This is one of the first PR to enable customizable loops in LitServe to facilitate interface for complex scenarios such as continuous batching with LLMs and diffusion models. 

The final goal is to enable users to provide custom Loop to `LitServer` to handle their own implementation of model inference cycle. 

```python
import litserve as ls

class MyLitAPI(ls.LitAPI):
    def decode_request(self, x):
        ...

    def predict(self, x):
        ...
    
    def encode_response(self, x):
        ...

    def  store_kv_cache(self, x):
        ...
    
    def get_kv_cache(self, x):
        ...

class CustomLoop(ls.Loop):
    def run(self):
        "handles the new hooks in LitAPI"
```

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/litserve/loops.py [hunk 1]
```diff
@@ -389,6 +389,76 @@ def run_batched_streaming_loop(
             response_queues[response_queue_id].put((uid, (e, LitAPIStatus.ERROR)))
 
 
+class _BaseLoop(ABC):
+    """A loop runs certain set of hooks implemented in LitAPI.
+
+    For a default loop, LitAPI must implement the following hooks:
+      - decode_request
```
Comment (Borda) [inline src/litserve/loops.py:396]: could this be a sphinx crosslink?

**Review by Borda — COMMENTED** [review]
can we add a test?

File: src/litserve/loops.py (hunk 1 above)
Comment (aniketmaurya) [reply src/litserve/loops.py:396]: we don't use sphinx docs here

**Comment by aniketmaurya** [comment]
@Borda the changes are already covered in the unit test. Will be adding more PRs to fully enable this feature :)

**Review by Borda — APPROVED** [review]

**Pull request merged**